src/dbnode/persist/fs/streaming_write.go

```diff
@@ -0,0 +1,259 @@
+	"github.com/m3db/m3/src/x/context"
+	"github.com/m3db/m3/src/x/ident"
+)
+
+// StreamingWriter writes int data fileset without intermediate buffering.
+// Writes must be lexicographically ordered by the id.
+type StreamingWriter interface {
+	Open() error
+	Write(ctx context.Context, encoder encoding.Encoder, id ident.ID, encodedTags []byte) error
+	Close() error
+}
+
+// StreamingWriterOptions in the options for the StreamingWriter.
+type StreamingWriterOptions struct {
+	Options             Options
+	NamespaceID         ident.ID
+	ShardID             uint32
+	BlockStart          time.Time
+	BlockSize           time.Duration
+	VolumeIndex         int
+	PlannedRecordsCount uint
+}
+
+type streamingWriter struct {
+	opts         StreamingWriterOptions
+	writer       *writer
+	writerOpts   DataWriterOpenOptions
+	data         []checked.Bytes
+	currIdx      int64
+	prevIDBytes  []byte
+	summaryEvery int64
+	bloomFilter  *bloom.BloomFilter
+	indexOffset  int64
+	summaries    int
+}
+
+func NewStreamingWriter(opts StreamingWriterOptions) (StreamingWriter, error) {
+	w, err := NewWriter(opts.Options)
+	if err != nil {
+		return nil, err
+	}
+
+	writerOpts := DataWriterOpenOptions{
+		BlockSize: opts.BlockSize,
+		Identifier: FileSetFileIdentifier{
+			Namespace:   opts.NamespaceID,
+			Shard:       opts.ShardID,
+			BlockStart:  opts.BlockStart,
+			VolumeIndex: opts.VolumeIndex,
+		},
+		FileSetType: persist.FileSetFlushType,
+	}
+
+	m, k := bloom.EstimateFalsePositiveRate(
+		opts.PlannedRecordsCount,
+		opts.Options.IndexBloomFilterFalsePositivePercent(),
+	)
+	bloomFilter := bloom.NewBloomFilter(m, k)
+
+	summariesApprox := float64(opts.PlannedRecordsCount) * opts.Options.IndexSummariesPercent()
+	summaryEvery := 0
+	if summariesApprox > 0 {
+		summaryEvery = int(math.Floor(float64(opts.PlannedRecordsCount) / summariesApprox))
+	}
+
+	return &streamingWriter{
+		opts:         opts,
+		writer:       w.(*writer),
+		writerOpts:   writerOpts,
+		summaryEvery: int64(summaryEvery),
+		data:         make([]checked.Bytes, 2),
+		bloomFilter:  bloomFilter,
+	}, nil
+}
+
+func (w *streamingWriter) Open() error {
+	if err := w.writer.Open(w.writerOpts); err != nil {
+		return err
+	}
+	w.indexOffset = 0
+	w.summaries = 0
+	w.prevIDBytes = nil
+	return nil
+}
+
+func (w *streamingWriter) Write(
+	ctx context.Context,
+	encoder encoding.Encoder,
+	id ident.ID,
+	encodedTags []byte,
+) error {
+	// Need to check if w.prevIDBytes != nil, otherwise we can never write an empty string ID
+	if w.prevIDBytes != nil && bytes.Compare(id.Bytes(), w.prevIDBytes) <= 0 {
+		return fmt.Errorf("ids must be written in lexicographic order, no duplicates, but got %s followed by %s", w.prevIDBytes, id)
+	}
+	w.prevIDBytes = append(w.prevIDBytes[:0], id.Bytes()...)
+
+	stream, ok := encoder.Stream(ctx)
+	if !ok {
+		// None of the datapoints passed the predicate.
+		return nil
+	}
+	defer stream.Finalize()
+	segment, err := stream.Segment()
+	if err != nil {
+		return err
+	}
```

Comment:
Wouldn't it be better just to do `segment := encoder.Discard()` here?

Reply:
+1

Reply:
Will change it.